crypto/pkcs7/pkcs7.c

```diff
@@ -1409,3 +1415,246 @@ int PKCS7_decrypt(PKCS7 *p7, EVP_PKEY *pkey, X509 *cert, BIO *data, int flags) {
+  if (si->auth_attr && sk_X509_ATTRIBUTE_num(si->auth_attr) != 0) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_INVALID_SIGNED_DATA_TYPE);
+    goto err;
+  }
+
+  EVP_PKEY *pkey;
+  if ((pkey = X509_get0_pubkey(signer)) == NULL) {
+    goto err;
+  }
+
+  ASN1_OCTET_STRING *data_body = si->enc_digest;
+  if (!EVP_VerifyFinal(mdc_copy, data_body->data, data_body->length, pkey)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_SIGNATURE_FAILURE);
+    goto err;
+  }
+
+  ret = 1;
+
+err:
+  EVP_MD_CTX_free(mdc_copy);
+  return ret;
+}
+
+int PKCS7_verify(PKCS7 *p7, STACK_OF(X509) *certs, X509_STORE *store,
+                 BIO *indata, BIO *out, int flags) {
+  GUARD_PTR(p7);
+  GUARD_PTR(store);
+  STACK_OF(X509) *signers = NULL, *untrusted = NULL;
+  X509_STORE_CTX *cert_ctx = NULL;
+  BIO *p7bio = NULL;
+  int ret = 0;
+
+  if (!PKCS7_type_is_signed(p7)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+    goto out;
+  }
+
+  // If |p7| is detached, caller must supply data via |indata|
+  if (PKCS7_is_detached(p7) && indata == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    goto out;
+  }
+
+  // We enforce OpenSSL's PKCS7_NO_DUAL_CONTENT flag in all cases for signed
+  if (PKCS7_type_is_signed(p7) && (!PKCS7_is_detached(p7) && indata)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_CONTENT_AND_DATA_PRESENT);
+    goto out;
+  }
+
+  STACK_OF(PKCS7_SIGNER_INFO) *sinfos = PKCS7_get_signer_info(p7);
+  if (sinfos == NULL || sk_PKCS7_SIGNER_INFO_num(sinfos) == 0UL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_SIGNATURES_ON_DATA);
+    goto out;
+  }
+
+  if ((signers = pkcs7_get0_signers(p7, certs, flags)) == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_SIGNERS);
+    goto out;
+  }
+
+  if (!(flags & PKCS7_NOVERIFY)) {
+    STACK_OF(X509) *included_certs = pkcs7_get0_certificates(p7);
+    if ((cert_ctx = X509_STORE_CTX_new()) == NULL ||
+        !pkcs7_x509_add_certs_new(&untrusted, certs) ||
+        !pkcs7_x509_add_certs_new(&untrusted, included_certs)) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+      goto out;
+    }
+
+    for (size_t k = 0; k < sk_X509_num(signers); k++) {
+      X509 *signer = sk_X509_value(signers, k);
+      if (!X509_STORE_CTX_init(cert_ctx, store, signer, untrusted)) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_X509_LIB);
+        goto out;
+      }
+      if (!X509_STORE_CTX_set_default(cert_ctx, "smime_sign")) {
+        goto out;
+      }
+      X509_STORE_CTX_set0_crls(cert_ctx, p7->d.sign->crl);
+    }
+    // NOTE: unlike most of our functions, |X509_verify_cert| can return <= 0
+    if (X509_verify_cert(cert_ctx) <= 0) {
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_CERTIFICATE_VERIFY_ERROR);
+      goto out;
+    }
+  }
+
+  // In copying data into out, we also read it through digest filters on |p7| to
+  // calculate digest for verification.
+  if ((p7bio = PKCS7_dataInit(p7, indata)) == NULL ||
+      (out && !pkcs7_bio_copy_content(p7bio, out))) {
```

Comment:
* Having a variable named `out` that is the same as the label we `goto` is confusing; can one of these be changed?
* Does `p7bio` need to be reset after the copy? Do we need to read from `p7bio` again after this?

Reply:
- yeah, in hindsight that is confusing. will change.
- we don't want to reset `p7bio` because we need to get its `EVP_MD_CTX` in `pkcs7_signature_verify`. the `pkcs7_bio_copy_content` call serves 2 purposes here -- 1. populate `outdata` BIO and more importantly 2. read data out through `p7bio`'s `BIO_f_md` to calculate the digest of `indata`'s contents.

Reply:
ah, but this highlights a gap! In OpenSSL, `PKCS7_verify`'s `out` parameter [is optional](https://docs.openssl.org/3.3/man3/PKCS7_verify/#name). same for us. however, because of this check on L1627 we'll never read out the contents of `indata` and populate `p7bio`'s digest.

we didn't catch this because we didn't have a test case eliding the `out` BIO. i'll fix this and add such a test case.